### PR TITLE
Add snapshot functionality for GlusterFS.

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by rootfs** [comment]
thanks
cc @tsmetana

File: snapshot/pkg/apis/crd/v1/types.go [hunk 1]
```diff
@@ -235,6 +241,9 @@ type VolumeSnapshotDataSource struct {
 	// kubelet's host machine and then exposed to the pod.
 	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
 	// +optional
+	//GlusterSnapshotSource represents a gluster snapshot resource
+	GlusterSnapshotVolume *GlusterVolumeSnapshotSource `json:"glusterSnapshotVolume,omitempty"`
```
Comment (tsmetana) [inline snapshot/pkg/apis/crd/v1/types.go]: It's not discoverable by the controller: you should also update the GetSupportedVolumeFromPVSpec and GetSupportedVolumeFromSnapshotDataSpec functions.

Comment (humblec) [reply snapshot/pkg/apis/crd/v1/types.go]: @tsmetana Thanks, Done. :)

File: snapshot/pkg/volume/gluster/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+func (h *glusterfsPlugin) SnapshotCreate(pv *v1.PersistentVolume, tags *map[string]string) (*crdv1.VolumeSnapshotDataSource, *[]crdv1.VolumeSnapshotCondition, error) {
+	spec := &pv.Spec
+	if spec == nil || spec.Glusterfs == nil {
+		return nil, nil, fmt.Errorf("invalid PV spec %v", spec)
+	}
+
+	volumePath := spec.Glusterfs.Path
+	snapshotName := volumePath + string(uuid.NewUUID())
+	cmd := exec.Command("sudo", glusterfsBinary, "snapshot", "create", snapshotName, volumePath)
```
Comment (tsmetana) [inline snapshot/pkg/volume/gluster/snapshot.go]: I don't think "sudo" is a right thing to do... The controller must run under an user who can use the glusterfs utilities and sudo should not be required. The controller should also work when run in container.

Comment (humblec) [reply snapshot/pkg/volume/gluster/snapshot.go]: @tsmetana sure, addressed the comment.

**Comment by tsmetana** [comment]
Looking good. Could you also please update the documentation and examples?

**Comment by humblec** [comment]
@tsmetana Done. The documentation is updated.  More improvements can be done by follow up PRs.

**Comment by tsmetana** [comment]
Thank you!

**Comment by humblec** [comment]
@wongma7 PTAL.

**Pull request merged**